ynr/apps/people/merging.py

```diff
@@ -0,0 +1,202 @@
+        self.source_person.loggedaction_set.update(person=self.dest_person)
+
+    def deep_merge_related_membership_objects(self, msource, mdest):
+        """
+        This method should update all related objects on msource, to point to
+        mdest. This only happens when two Memberships with
+        (Person ID, Ballot Paper) would exist when merging two people. This
+        case shoudl be very rare, as it would only happen when the same person
+        got added twice to the same ballot, just with duplicate Person models.
+
+        We should be able to safely delete msource at the end of this method.
+
+        Assume that the non-related field data on mdest is the correct data,
+        e.g `party_list_position`.
+
+        Also assume that `Party` is "more correct" in mdest.
+        """
+        if hasattr(msource, "result"):
+            if hasattr(mdest, "result"):
+                raise InvalidMergeError(
+                    "Trying to merge two Memberships with results"
+                )
+            else:
+                msource.result.membership = mdest
+                mdest.result.save()
+
+        self.safe_delete(msource)
+
+    def merge_memberships(self):
+        for membership in self.source_person.memberships.all():
+            existing_membership_qs = self.dest_person.memberships.filter(
+                post_election=membership.post_election
+            )
+            if existing_membership_qs.exists():
+                # This is a duplicate membership, so we need to merge the
+                # related objects
+                self.deep_merge_related_membership_objects(
+                    membership, existing_membership_qs.get()
+                )
+            else:
+                membership.person = self.dest_person
+                membership.save()
+        # Check that we've not causes duplicate (membership, election) pairs.
+        # We need to do this manually because we can't add a DB constraint
+        # spanning the three tables (Membership->Ballot->Election)
+        person_elections = {
+            mem.post_election.election_id
+            for mem in self.dest_person.memberships.select_related(
+                "post_election"
+            )
+        }
+        if len(person_elections) != self.dest_person.memberships.count():
+            raise InvalidMergeError(
+                "Merging would cause this person to be standing more than once in the same election"
+            )
+
+    def setup_redirect(self):
+        # Create a redirect from the old person to the new person:
+        PersonRedirect.objects.create(
+            old_person_id=self.source_person.pk,
+            new_person_id=self.dest_person.pk,
+        )
+
+    def merge(self, delete=True):
+        """
+        Do everything we need in order to merge two people.
+
+        At the end of this, `self.source_person` should have no related objects
+        left, as checked by `self.check_safe_to_delete()`
```

Comment:
This class doesn't have a method called `check_safe_to_delete()`. Should this be `safe_delete()` ?